File: android/app/src/main/kotlin/com/safeguard/app/engine/shield/ContentShield.kt
package com.safeguard.app.engine.shield

import com.safeguard.app.engine.ai.ClassificationResult
import com.safeguard.app.engine.ai.ClassificationStatus
import com.safeguard.app.engine.ai.ContentKind
import com.safeguard.app.engine.ai.DecisionPolicy
import com.safeguard.app.engine.ai.FinalAction
import com.safeguard.app.engine.ai.RuleSignal
import com.safeguard.app.engine.explain.DecisionExplainer
import com.safeguard.app.engine.rules.Category

/** The user's shield settings. Off until the user turns it on (it needs an accessibility permission). */
data class ShieldSettings(
    val enabled: Boolean = false,
    /** Supported-app keys the user switched off; every other supported app is inspected. */
    val disabledApps: Set<String> = emptySet(),
) {
    fun isAppEnabled(app: SupportedApp) = app.key !in disabledApps

    fun enabledApps(): List<SupportedApp> = SupportedApps.all.filter(::isAppEnabled)

    /** Package ids the accessibility service should receive events from (empty = none). */
    fun monitoredPackages(): List<String> = if (enabled) enabledApps().flatMap { it.packages }.sorted() else emptyList()

    /** True if [next] inspects less (the UI requires the PIN). */
    fun isLoosenedBy(next: ShieldSettings): Boolean =
        (enabled && !next.enabled) || (enabled && (next.disabledApps - disabledApps).isNotEmpty())
}

enum class ShieldState(val id: String) {
    /** The user hasn't turned the shield on. */
    OFF("off"),

    /** Text and image classification both running. */
    ACTIVE("active"),

    /** Running, but not every content type is covered (e.g. no image model). */
    PARTIAL("partial"),

    /** Turned on, but nothing is actually being inspected. */
    UNAVAILABLE("unavailable"),
}

/** Why the shield isn't fully active. Stable ids; the UI translates them. */
enum class ShieldIssue(val id: String) {
    PROTECTION_OFF("protection_off"),
    AI_DISABLED("ai_disabled"),
    ACCESSIBILITY_OFF("accessibility_off"),
    ACCESSIBILITY_UNAVAILABLE("accessibility_unavailable"),
    NO_APPS("no_apps"),
    TEXT_MODEL_UNAVAILABLE("text_model_unavailable"),
    IMAGE_MODEL_UNAVAILABLE("image_model_unavailable"),
    INFERENCE_SLOW("inference_slow"),
}

data class ShieldStatus(
    val state: ShieldState,
    val issues: List<ShieldIssue>,
    val textActive: Boolean,
    val imageActive: Boolean,
)

/**
 * The shield's real state. It never reports ACTIVE unless both text and
 * image classification can run right now, and never anything but OFF or
 * UNAVAILABLE when nothing is inspected.
 */
object ShieldStatusResolver {
    fun resolve(
        settings: ShieldSettings,
        /** Protection on and not paused. */
        protectionActive: Boolean,
        aiEnabled: Boolean,
        accessibilityEnabled: Boolean,
        accessibilityAvailable: Boolean,
        textModelAvailable: Boolean,
        imageModelAvailable: Boolean,
        textSlow: Boolean = false,
        imageSlow: Boolean = false,
    ): ShieldStatus {
        if (!settings.enabled) return ShieldStatus(ShieldState.OFF, emptyList(), false, false)
        val blocking = buildList {
            if (!protectionActive) add(ShieldIssue.PROTECTION_OFF)
            if (!aiEnabled) add(ShieldIssue.AI_DISABLED)
            if (!accessibilityAvailable) add(ShieldIssue.ACCESSIBILITY_UNAVAILABLE)
            else if (!accessibilityEnabled) add(ShieldIssue.ACCESSIBILITY_OFF)
            if (settings.enabledApps().isEmpty()) add(ShieldIssue.NO_APPS)
        }
        val partial = buildList {
            if (!textModelAvailable) add(ShieldIssue.TEXT_MODEL_UNAVAILABLE)
            if (!imageModelAvailable) add(ShieldIssue.IMAGE_MODEL_UNAVAILABLE)
            if ((textModelAvailable && textSlow) || (imageModelAvailable && imageSlow)) add(ShieldIssue.INFERENCE_SLOW)
        }
        val text = blocking.isEmpty() && textModelAvailable && !textSlow
        val image = blocking.isEmpty() && imageModelAvailable && !imageSlow
        val state = when {
            !text && !image -> ShieldState.UNAVAILABLE
            partial.isEmpty() -> ShieldState.ACTIVE
            else -> ShieldState.PARTIAL
        }
        return ShieldStatus(state, blocking + partial, text, image)
    }
}

/**
 * Inference that takes too long is treated as a failure: after
 * [maxConsecutive] slow runs the content kind is paused for [pauseMs]
 * (then retried), so a slow device can't be kept busy in a loop. A running
 * inference can't be interrupted; the pause stops new ones.
 */
class InferenceWatchdog(
    private val limitMs: Long,
    private val maxConsecutive: Int = 3,
    private val pauseMs: Long = 60_000,
) {
    private var consecutive = 0
    private var pausedUntil = Long.MIN_VALUE

    @Synchronized
    fun canRun(now: Long) = now >= pausedUntil

    @Synchronized
    fun isSlow(now: Long) = now < pausedUntil

    /** Records one run; returns true if it counts as a timeout. */
    @Synchronized
    fun record(now: Long, elapsedMs: Long): Boolean {
        if (elapsedMs <= limitMs) {
            consecutive = 0
            return false
        }
        consecutive++
        if (consecutive >= maxConsecutive) {
            pausedUntil = now + pauseMs
            consecutive = 0
        }
        return true
    }
}

/** What happened with one sample. */
sealed interface ShieldOutcome {
    /** Not inspected (shield off, app not enabled, throttled, same screen…). */
    data object Skipped : ShieldOutcome

    /** Classified; the policy engine allowed it (or said UNKNOWN). */
    data class Allowed(val decision: ShieldDecision) : ShieldOutcome

    /** Classified as risky once; waiting for a confirming sample. */
    data class Pending(val decision: ShieldDecision) : ShieldOutcome

    /** Blocked: the caller hides the content. */
    data class Blocked(val event: ShieldEvent) : ShieldOutcome
}

/**
 * Everything the log keeps about a shield block: no text, no image, no
 * hash of either. [packageName] is the app (as for protected apps).
 */
data class ShieldEvent(
    val timestamp: Long,
    val packageName: String,
    val kind: ContentKind,
    val label: AiLabel,
    val category: Category,
    /** Confidence rounded down to 10 % steps. */
    val confidenceBucket: Int,
    val modelVersion: String,
) {
    /** Stored as the event's rule type: `ai_shield:<kind>:<label>:<model>`. */
    val ruleType: String get() = "$RULE_TYPE_PREFIX${kind.name.lowercase()}:${label.id}:$modelVersion"

    companion object {
        const val RULE_TYPE_PREFIX = DecisionExplainer.SHIELD_RULE_TYPE_PREFIX
    }
}

/**
 * AI Content Shield orchestration, independent of Android:
 *
 * sample (text snapshot or frame) → gate (throttle, debounce, same-screen
 * skip) → classify on device (with a watchdog) → temporal confirmation →
 * existing policy engine → BLOCK / ALLOW / UNKNOWN.
 *
 * Only packages in [SupportedApps] that the user left enabled are ever
 * considered, and nothing runs while protection or the shield is off.
 * Inputs are never stored; only [ShieldEvent] metadata is reported for
 * blocks.
 */
class ContentShieldEngine(
    private val classifyText: (String) -> ClassificationResult,
    private val image: ShieldImageClassifier,
    private val policy: () -> DecisionPolicy,
    private val settings: () -> ShieldSettings,
    /** Protection on, not paused. */
    private val protectionActive: () -> Boolean,
    private val clock: () -> Long = System::currentTimeMillis,
    private val textWatchdog: InferenceWatchdog = InferenceWatchdog(limitMs = 250),
    private val imageWatchdog: InferenceWatchdog = InferenceWatchdog(limitMs = 1_500),
    /** After a block, the same app isn't blocked again for this long (the user is sent home). */
    private val blockCooldownMs: Long = 3_000,
) {
    private val textGate = FrameGate(minIntervalMs = 1_000, settleMs = 400, maxWaitMs = 2_500, sameScreenBits = 0)
    private val imageGate = FrameGate()
    private val textConfirm = TemporalConfirmer()
    private val imageConfirm = TemporalConfirmer()
    @Volatile private var foreground: String? = null
    @Volatile private var lastTextHash: Long? = null
    @Volatile private var textPending = false
    private val lastBlockAt = HashMap<String, Long>()

    val textSlow: Boolean get() = textWatchdog.isSlow(clock())
    val imageSlow: Boolean get() = imageWatchdog.isSlow(clock())

    /** Whether content of [pkg] may be inspected right now. */
    fun isActiveFor(pkg: String?): Boolean {
        val app = SupportedApps.forPackage(pkg) ?: return false
        val s = settings()
        return s.enabled && s.isAppEnabled(app) && protectionActive() && policy().ai.enabled
    }

    /** The foreground app changed: evidence from the previous app is dropped. */
    fun onForeground(pkg: String?) {
        if (pkg == foreground) return
        foreground = pkg
        lastTextHash = null
        textPending = false
        textGate.reset(); imageGate.reset()
        textConfirm.reset(); imageConfirm.reset()
    }

    /** Scrolling or other content changes (debounces sampling). */
    fun onContentChanged() {
        val now = clock()
        textGate.onContentChanged(now)
        imageGate.onContentChanged(now)
    }

    /** Whether a text snapshot should be taken now (call before walking the node tree). */
    fun wantsText(pkg: String?): Boolean =
        isActiveFor(pkg) && textWatchdog.canRun(clock()) && textGate.shouldSample(clock(), null, true)

    fun onText(pkg: String, text: String): ShieldOutcome {
        if (!isActiveFor(pkg) || text.isBlank()) return ShieldOutcome.Skipped
        // Same text as last time and nothing to confirm: nothing new to learn.
        val hash = FrameHash.text(text)
        if (hash == lastTextHash && !textPending) return ShieldOutcome.Skipped
        lastTextHash = hash
        val start = clock()
        val result = try {
            classifyText(text)
        } catch (e: OutOfMemoryError) {
            ClassificationResult(ClassificationStatus.UNAVAILABLE)
        } catch (e: Exception) {
            ClassificationResult(ClassificationStatus.UNAVAILABLE)
        }
        textWatchdog.record(clock(), clock() - start)
        return decide(pkg, AiClassification.fromText(result), textConfirm, textGate).also {
            textPending = it is ShieldOutcome.Pending
        }
    }

    /** [frame] is read in place and must be released by the caller afterwards. */
    fun onFrame(pkg: String, frame: RgbaFrame): ShieldOutcome {
        if (!isActiveFor(pkg) || !image.isUsable) return ShieldOutcome.Skipped
        val now = clock()
        if (!imageWatchdog.canRun(now)) return ShieldOutcome.Skipped
        if (!imageGate.shouldSample(now, FrameHash.dHash(frame), true)) return ShieldOutcome.Skipped
        val c = image.classify(frame)
        imageWatchdog.record(clock(), clock() - now)
        return decide(pkg, c, imageConfirm, imageGate)
    }

    /** Frees the image model (protection or shield stopped). */
    fun release() {
        image.close()
        onForeground(null)
    }

    private fun decide(pkg: String, c: AiClassification, confirm: TemporalConfirmer, gate: FrameGate): ShieldOutcome {
        val p = policy()
        val now = clock()
        val evidence = confirm.add(now, ShieldScores.toResult(c, ShieldScores.blockSuggestive(p.ai.mode)))
        val d = ShieldPolicy.decide(RuleSignal.None, evidence, c, p)
        val pending = evidence.status == ClassificationStatus.UNCERTAIN && c.status == ClassificationStatus.OK
        gate.onResult(pending)
        if (d.decision.action != FinalAction.BLOCK) {
            return if (pending) ShieldOutcome.Pending(d) else ShieldOutcome.Allowed(d)
        }
        synchronized(lastBlockAt) {
            val last = lastBlockAt[pkg]
            // Already handled: the user was just sent away from this content.
            if (last != null && now - last < blockCooldownMs) return ShieldOutcome.Skipped
            lastBlockAt[pkg] = now
        }
        confirm.reset()
        val label = if (c.label.isRisk && c.label.category == d.decision.category) c.label else AiLabel.of(d.decision.category)
        return ShieldOutcome.Blocked(
            ShieldEvent(
                timestamp = now,
                packageName = pkg,
                kind = c.kind,
                label = label,
                category = d.decision.category,
                confidenceBucket = confidenceBucket(d.decision.confidence),
                modelVersion = c.modelVersion ?: "unknown",
            ),
        )
    }
}

File: lib/features/protection/domain/content_shield.dart
import '../../../core/i18n/i18n.dart';
import 'protection.dart';

/// AI Content Shield state, as reported by the native side. Mirrors
/// `ShieldState` in engine/shield/ContentShield.kt.
enum ShieldState {
  /// The user hasn't turned the shield on.
  off('off'),

  /// Text and image classification both running.
  active('active'),

  /// Running, but not every content type is covered (e.g. no image model).
  partial('partial'),

  /// Turned on, but nothing is actually being inspected.
  unavailable('unavailable');

  const ShieldState(this.id);
  final String id;

  static ShieldState fromId(Object? id) {
    for (final s in values) {
      if (s.id == id) return s;
    }
    return unavailable;
  }

  String get label => switch (this) {
    off => tr('متوقف', 'Off'),
    active => tr('يعمل', 'Active'),
    partial => tr('يعمل جزئيًا', 'Partially active'),
    unavailable => tr('غير متاح', 'Unavailable'),
  };
}

/// Why the shield isn't fully active. Mirrors `ShieldIssue`.
enum ShieldIssue {
  protectionOff('protection_off'),
  aiDisabled('ai_disabled'),
  accessibilityOff('accessibility_off'),
  accessibilityUnavailable('accessibility_unavailable'),
  noApps('no_apps'),
  textModelUnavailable('text_model_unavailable'),
  imageModelUnavailable('image_model_unavailable'),
  inferenceSlow('inference_slow');

  const ShieldIssue(this.id);
  final String id;

  static ShieldIssue? fromId(Object? id) {
    for (final s in values) {
      if (s.id == id) return s;
    }
    return null;
  }

  String get message => switch (this) {
    protectionOff => tr(
      'الحماية متوقفة أو مؤقتًا معطّلة، فلا يُفحص أي محتوى.',
      'Protection is off or paused, so no content is checked.',
    ),
    aiDisabled => tr(
      'الحماية الذكية متوقفة في إعداداتك.',
      'AI protection is turned off in your settings.',
    ),
    accessibilityOff => tr(
      'خدمة «درع المحتوى الذكي» غير مفعّلة في إعدادات تسهيل الاستخدام.',
      'The “AI Content Shield” service is off in Accessibility settings.',
    ),
    accessibilityUnavailable => tr(
      'هذا الجهاز لا يسمح بخدمات تسهيل الاستخدام لـ SafeGuard.',
      "This device doesn't allow SafeGuard's accessibility services.",
    ),
    noApps => tr(
      'لم يبقَ أي تطبيق مدعوم مفعّلًا.',
      'No supported app is left on.',
    ),
    textModelUnavailable => tr(
      'نموذج النصوص غير متاح حاليًا.',
      'The text model is unavailable right now.',
    ),
    imageModelUnavailable => tr(
      'لا يوجد نموذج صور في هذا الإصدار: الصور والفيديو لا تُفحص.',
      'No image model in this version: images and video are not checked.',
    ),
    inferenceSlow => tr(
      'التحليل بطيء على هذا الجهاز، فأُوقف مؤقتًا.',
      'Analysis is too slow on this device, so it is paused for now.',
    ),
  };
}

/// Why image classification can't run. Mirrors `ImageModelState`.
enum ImageModelState {
  notBundled('not_bundled'),
  noRuntime('no_runtime'),
  corrupted('corrupted'),
  loadFailed('load_failed'),
  outOfMemory('out_of_memory'),
  ready('ready'),
  notLoaded('not_loaded');

  const ImageModelState(this.id);
  final String id;

  static ImageModelState fromId(Object? id) {
    for (final s in values) {
      if (s.id == id) return s;
    }
    return notBundled;
  }

  bool get usable => this == ready || this == notLoaded;

  String get label => switch (this) {
    notBundled => tr(
      'غير مضمَّن في هذا الإصدار',
      'Not included in this version',
    ),
    noRuntime => tr('لا يوجد محرك تشغيل', 'No runtime available'),
    corrupted => tr('الملف تالف أو غير موثوق', 'File damaged or not trusted'),
    loadFailed => tr('تعذّر التحميل', "Couldn't load"),
    outOfMemory => tr('الذاكرة غير كافية', 'Not enough memory'),
    ready => tr('جاهز', 'Ready'),
    notLoaded => tr('جاهز (يُحمَّل عند الحاجة)', 'Ready (loads when needed)'),
  };
}

/// What SafeGuard expects to be able to read in an app. Mirrors `SupportLevel`.
enum ShieldSupportLevel {
  text('text'),
  textLimited('text_limited');

  const ShieldSupportLevel(this.id);
  final String id;

  static ShieldSupportLevel fromId(Object? id) =>
      id == 'text' ? text : textLimited;

  String get label => switch (this) {
    text => tr('نصوص الصفحات والمنشورات', 'Page and post text'),
    textLimited => tr(
      'نصوص محدودة (عناوين ووصف)',
      'Limited text (titles and captions)',
    ),
  };
}

/// Known reasons content may be missed. Mirrors `ShieldLimitation`.
enum ShieldLimitation {
  imagesNeedModel('images_need_model'),
  mostlyVideo('mostly_video'),
  customRendering('custom_rendering'),
  secureSurfaces('secure_surfaces'),
  uiChanges('ui_changes'),
  privateMessagesVisible('private_messages_visible');

  const ShieldLimitation(this.id);
  final String id;

  static ShieldLimitation? fromId(Object? id) {
    for (final s in values) {
      if (s.id == id) return s;
    }
    return null;
  }

  String get message => switch (this) {
    imagesNeedModel => tr(
      'الصور والفيديو لا تُفحص (لا يوجد نموذج صور).',
      "Photos and video aren't checked (no image model).",
    ),
    mostlyVideo => tr(
      'معظم المحتوى فيديو؛ يُفحص نص العناوين والوصف فقط.',
      'Mostly video; only titles and captions are checked.',
    ),
    customRendering => tr(
      'قد لا يُظهر التطبيق بعض النصوص لخدمات تسهيل الاستخدام.',
      'The app may not expose some text to accessibility services.',
    ),
    secureSurfaces => tr(
      'الشاشات المحمية في التطبيق لا يمكن قراءتها.',
      "Screens the app marks as secure can't be read.",
    ),
    uiChanges => tr(
      'قد تغيّر تحديثات التطبيق ما يمكن قراءته في أي وقت.',
      'App updates can change what can be read at any time.',
    ),
    privateMessagesVisible => tr(
      'الرسائل الظاهرة على الشاشة تُفحص في الذاكرة مثل أي نص، ولا تُخزَّن.',
      "Messages shown on screen are checked in memory like any text, and aren't stored.",
    ),
  };
}

class ShieldApp {
  const ShieldApp({
    required this.key,
    required this.name,
    required this.packages,
    required this.level,
    required this.limitations,
    required this.enabled,
    required this.installed,
    required this.verifiedOnDevice,
  });

  final String key;
  final String name;
  final List<String> packages;
  final ShieldSupportLevel level;
  final List<ShieldLimitation> limitations;
  final bool enabled;
  final bool installed;

  /// True only after a real-device test run (docs/FINAL_AI_TESTING.md).
  final bool verifiedOnDevice;

  static ShieldApp? fromMap(Object? raw) {
    if (raw is! Map) return null;
    final key = raw['key'];
    final name = raw['name'];
    if (key is! String || name is! String) return null;
    return ShieldApp(
      key: key,
      name: name,
      packages: [
        for (final p in (raw['packages'] as List?) ?? const [])
          if (p is String) p,
      ],
      level: ShieldSupportLevel.fromId(raw['level']),
      limitations: [
        for (final l in (raw['limitations'] as List?) ?? const [])
          ?ShieldLimitation.fromId(l),
      ],
      enabled: raw['enabled'] != false,
      installed: raw['installed'] == true,
      verifiedOnDevice: raw['verifiedOnDevice'] == true,
    );
  }
}

class ShieldStatus {
  const ShieldStatus({
    required this.enabled,
    required this.state,
    required this.issues,
    required this.textActive,
    required this.imageActive,
    required this.accessibility,
    required this.textModel,
    required this.textModelAvailable,
    required this.imageModelState,
    required this.apps,
  });

  /// The user turned the shield on.
  final bool enabled;
  final ShieldState state;
  final List<ShieldIssue> issues;
  final bool textActive;
  final bool imageActive;

  /// State of the shield's own accessibility service.
  final AccessibilityStatus accessibility;
  final String textModel;
  final bool textModelAvailable;
  final ImageModelState imageModelState;
  final List<ShieldApp> apps;

  /// Shown on platforms without the native layer.
  static const unsupported = ShieldStatus(
    enabled: false,
    state: ShieldState.unavailable,
    issues: [],
    textActive: false,
    imageActive: false,
    accessibility: AccessibilityStatus.unsupported,
    textModel: 'sg-text-1',
    textModelAvailable: false,
    imageModelState: ImageModelState.notBundled,
    apps: [],
  );

  bool get isChecking => textActive || imageActive;

  static ShieldStatus fromMap(Map<Object?, Object?> m) => ShieldStatus(
    enabled: m['enabled'] == true,
    state: ShieldState.fromId(m['state']),
    issues: [
      for (final i in (m['issues'] as List?) ?? const [])
        ?ShieldIssue.fromId(i),
    ],
    textActive: m['textActive'] == true,
    imageActive: m['imageActive'] == true,
    accessibility: AccessibilityStatus.fromId(m['accessibility']),
    textModel: m['textModel'] is String ? m['textModel']! as String : '',
    textModelAvailable: m['textModelAvailable'] == true,
    imageModelState: ImageModelState.fromId(m['imageModelState']),
    apps: [
      for (final a in (m['apps'] as List?) ?? const []) ?ShieldApp.fromMap(a),
    ],
  );

  ShieldStatus copyWith({bool? enabled, List<ShieldApp>? apps}) => ShieldStatus(
    enabled: enabled ?? this.enabled,
    state: state,
    issues: issues,
    textActive: textActive,
    imageActive: imageActive,
    accessibility: accessibility,
    textModel: textModel,
    textModelAvailable: textModelAvailable,
    imageModelState: imageModelState,
    apps: apps ?? this.apps,
  );
}
